File: user/psava/Mdecibel.c
/* Decibel */
#include <rsf.h>
#include <math.h>

/*------------------------------------------------------------*/
int main(int argc, char* argv[])
{
    bool verb;     /* verbosity flag */
    bool inv;

    sf_file  Fin, Fou; /* I/O files */

    sf_axis a1; /* cube axes */
    int     n,nbuf,ibuf;

    float aref;
    float *trace;

    sf_init(argc,argv);

    if(! sf_getbool("verb",&verb)) verb=false; /* verbosity flag */
    if(! sf_getfloat("aref",&aref)) aref=1.0;   /* reference amplitude */
    if(! sf_getbool("inv",&inv)) inv=false; /* inverse transform */

    Fin = sf_input ("in" );
    Fou = sf_output("out");

    a1=sf_iaxa(Fin,1); if(verb) sf_raxa(a1); 

    nbuf = sf_bufsiz(Fin)/sizeof(float);
    trace = sf_floatalloc (nbuf);

    for (n = sf_filesize(Fin); n > 0; n -= nbuf) {
        if (nbuf > n) nbuf=n;

        sf_floatread(trace,nbuf,Fin);

	if(inv) {
	    for(ibuf=0; ibuf<nbuf; ibuf++)
		trace[ibuf] = aref * pow( 10,0.05*trace[ibuf]);
	} else {
	    for(ibuf=0; ibuf<nbuf; ibuf++)
		trace[ibuf] = 20. * log10( fabs(trace[ibuf]/aref) );
	}

        sf_floatwrite(trace,nbuf,Fou);

    }

    free(trace);

}

File: user/psava/Mdecibel.py
#!/usr/bin/env python
'''
convert amplitude to decibels
'''
import rsf.api as rsf
import numpy as np
import sys,math

par = rsf.Par()

verb = par.bool('verb',False) # verbosity flag
aref = par.float('aref',1.0)  # reference amplitude
inv = par.bool('inv',False) # invere transform (decibel to amplitude)

Fin = rsf.Input()         # input
n1 = Fin.int("n1")
o1 = Fin.float("o1")
d1 = Fin.float("d1")

nn = Fin.size(1); # number of traces

Fou = rsf.Output()        # output 
Fou.put("n1",n1)
Fou.put("o1",o1)
Fou.put('d1',d1)

din = np.zeros(n1,'f') 
dou = np.zeros(n1,'f')

for ii in range(nn):
    Fin.read(din)

    if inv:
        for i1 in range(n1):
            dou[i1] = aref * pow(10,0.05*din[i1])
    else:
        for i1 in range(n1):
            if math.fabs(din[i1])>1e-6:
                dou[i1] = 20 * math.log10( math.fabs(din[i1]/aref) )
            else:
                dou[i1] = -120

    Fou.write(dou)
    
# ------------------------------------------------------------
Fin.close()
Fou.close()

